src/metabase/upload.clj

```diff
@@ -601,10 +601,90 @@
   [db table]
   (nil? (can-update-error db table)))
 
+(defn- can-delete-error
+  "Returns an ExceptionInfo object if the user cannot delete the given upload. Returns nil otherwise."
+  [table]
+  (cond
+    (not (:is_upload table))
+    (ex-info (tru "The table must be an uploaded table.")
+             {:status-code 422})
+
+    (not (mi/can-write? table))
+    (ex-info (tru "You don''t have permissions to do that.")
+             {:status-code 403})))
+
+(defn- check-can-delete
+  "Throws an error if the given table is not an upload, or if the user does not have permission to delete it."
+  [table]
+  ;; For now anyone that can update a table is allowed to delete it.
+  (when-let [error (can-delete-error table)]
+    (throw error)))
+
+(defn- table-exists?
+  "Check whether the give table exists within the customer database."
+  [driver database table]
+  (try
+    (driver/describe-table driver (:id database) table)
```

Comment:
I'd rather not dig down into sql-jdbc internals, as this will need to generalize to non-SQL drivers soon.

I'm aware this is not a perfect fit, but it does work for the current drivers and will be easy to generalize as we add BigQuery etc. I wonder now whether there isn't a way to use QP instead though, e.g. run an MBQL equivalent of `select 1 from {table} limit 1`? I'd hope the error message for something like is consistent across drivers.

Reply:
I don't think we need this function, see https://github.com/metabase/metabase/pull/41293/files#r1568680464

Reply:
Also for future reference, `driver/describe-table` is not supposed to be used for drivers that support the `:describe-fields` feature like redshift. They should use `driver/describe-fields` instead. I will make this clearer in the docstrings and also enforce it in an upcoming PR.

Reply:
Thanks for the pointer, and that docstring update soungs great. I guess you could even add a pre-condition, although I'm not sure they get elided in production.

Inclined to agree that the logging isn't worth all this hackery now that we don't need to rollback any card archival, I've removed it.